File: lib/features/saliks/data/area_repository.dart
import 'package:flutter_riverpod/flutter_riverpod.dart';
import 'package:uuid/uuid.dart';

import '../../../core/data/reference_data.dart';
import '../../../core/database/app_database.dart';
import '../domain/entities/area.dart';
import '../domain/entities/bazam.dart';

final areaRepositoryProvider = Provider<AreaRepository>((ref) {
  return AreaRepository(
    ref.watch(appDatabaseProvider),
  );
});

class AreaRepository {
  AreaRepository(this._db);

  final AppDatabase _db;
  final _uuid = const Uuid();

  static Area _fromRow(LocalArea r) {
    final bazamId = r.bazamId.trim().isEmpty ? kDefaultBazamId : r.bazamId;
    return Area(
      areaId: r.areaId,
      areaName: r.areaName,
      bazamId: bazamId,
    );
  }

  Stream<List<Area>> watchAreas() {
    return _db.watchAllAreas().map((rows) {
      final areas = rows
          .where((r) => r.syncStatus != pendingDelete && r.syncStatus != aliasSynced)
          .map(_fromRow)
          .toList()
        ..sort((a, b) => a.areaName.compareTo(b.areaName));
      if (areas.isNotEmpty) return areas;
      return [...kAreas]
        ..sort((a, b) => a.areaName.compareTo(b.areaName));
    });
  }

  Stream<List<Bazam>> watchBazams() {
    return _db.watchAllBazams().map((rows) {
      final bazams = rows
          .where((r) => r.syncStatus != pendingDelete)
          .map(
            (r) => Bazam(
              bazamId: r.bazamId,
              bazamName: r.bazamName,
            ),
          )
          .toList()
        ..sort((a, b) => a.bazamName.compareTo(b.bazamName));
      if (bazams.isNotEmpty) return bazams;
      return [...kBazams];
    });
  }

  Future<Bazam?> resolveBazam(String bazamId) async {
    if (bazamId.isEmpty) return null;
    final canonical = findBazam(bazamId);
    if (canonical != null) return canonical;

    final row = await _db.getBazamById(bazamId);
    if (row != null && row.syncStatus != pendingDelete) {
      return Bazam(bazamId: row.bazamId, bazamName: row.bazamName);
    }
    return null;
  }

  Future<Area?> resolveArea(String areaId) async {
    if (areaId.isEmpty) return null;

    final canonical = findArea(areaId);
    if (canonical != null) return canonical;

    final row = await _db.getAreaById(areaId);
    if (row != null && row.syncStatus != pendingDelete) {
      return _fromRow(row);
    }

    return null;
  }

  Future<List<Area>> _allAreasLocal() async {
    final rows = await _db.watchAllAreas().first;
    if (rows.isEmpty) return kAreas;
    return rows
        .where((r) => r.syncStatus != pendingDelete && r.syncStatus != aliasSynced)
        .map(_fromRow)
        .toList();
  }

  Future<List<Area>> areasForBazam(String bazamId) async {
    final id = bazamId.trim().isEmpty ? kDefaultBazamId : bazamId.trim();
    final areas = await _allAreasLocal();
    return areas.where((a) => a.bazamId == id).toList()
      ..sort((a, b) => a.areaName.compareTo(b.areaName));
  }

  Future<Area?> findAreaByName(String name) async {
    final normalized = name.trim().toLowerCase();
    if (normalized.isEmpty) return null;

    for (final area in await _allAreasLocal()) {
      if (area.areaName.trim().toLowerCase() == normalized) {
        return area;
      }
    }
    return null;
  }

  Future<Area> createArea({
    required String name,
    String bazamId = kDefaultBazamId,
  }) async {
    final existing = await findAreaByName(name);
    if (existing != null) return existing;

    final id = _uuid.v4();
    final area = Area(
      areaId: id,
      areaName: name.trim(),
      bazamId: bazamId.trim().isEmpty ? kDefaultBazamId : bazamId.trim(),
    );
    await _db.upsertArea(areaToCompanion(area, syncStatus: pendingCreate));
    await _db.enqueueSync(
      collection: 'areas',
      operation: 'create',
      docId: id,
      payload: area.toMap(),
    );
    return area;
  }
}

File: app/src/main/java/com/example/salik_management_system/features/saliks/data/repository/AreaRepository.kt
package com.example.salik_management_system.features.saliks.data.repository

import com.example.salik_management_system.core.data.kAreas
import com.example.salik_management_system.core.data.kBazams
import com.example.salik_management_system.core.data.findArea
import com.example.salik_management_system.core.data.findBazam
import com.example.salik_management_system.core.database.LocalAreaDao
import com.example.salik_management_system.core.database.LocalBazamDao
import com.example.salik_management_system.core.database.SyncStatus
import com.example.salik_management_system.core.database.enqueueSync
import com.example.salik_management_system.features.saliks.data.mapper.toDomain
import com.example.salik_management_system.features.saliks.data.mapper.toEntity
import com.example.salik_management_system.features.saliks.domain.model.Area
import com.example.salik_management_system.features.saliks.domain.model.Bazam
import com.example.salik_management_system.features.saliks.domain.model.kDefaultBazamId
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.map
import java.util.UUID
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class AreaRepository @Inject constructor(
    private val areaDao: LocalAreaDao,
    private val bazamDao: LocalBazamDao,
    private val syncQueueDao: com.example.salik_management_system.core.database.SyncQueueDao,
) {
    fun watchAreas(): Flow<List<Area>> =
        areaDao.watchAll().map { rows ->
            val areas = rows
                .filter {
                    it.syncStatus != SyncStatus.pendingDelete &&
                        it.syncStatus != SyncStatus.alias
                }
                .map { it.toDomain() }
                .sortedBy { it.areaName.lowercase() }
            if (areas.isNotEmpty()) areas else kAreas.sortedBy { it.areaName.lowercase() }
        }

    fun watchBazams(): Flow<List<Bazam>> =
        bazamDao.watchAll().map { rows ->
            val bazams = rows
                .filter { it.syncStatus != SyncStatus.pendingDelete }
                .map { it.toDomain() }
                .sortedBy { it.bazamName.lowercase() }
            if (bazams.isNotEmpty()) bazams else kBazams
        }

    suspend fun resolveBazam(bazamId: String): Bazam? {
        if (bazamId.isEmpty()) return null
        findBazam(bazamId)?.let { return it }
        val row = bazamDao.getById(bazamId) ?: return null
        if (row.syncStatus == SyncStatus.pendingDelete) return null
        return row.toDomain()
    }

    suspend fun resolveArea(areaId: String): Area? {
        if (areaId.isEmpty()) return null
        findArea(areaId)?.let { return it }
        val row = areaDao.getById(areaId) ?: return null
        if (row.syncStatus == SyncStatus.pendingDelete) return null
        return row.toDomain()
    }

    suspend fun areasForBazam(bazamId: String): List<Area> {
        val id = bazamId.trim().ifEmpty { kDefaultBazamId }
        return allAreasLocal()
            .filter { it.bazamId == id }
            .sortedBy { it.areaName.lowercase() }
    }

    suspend fun findAreaByName(name: String): Area? {
        val normalized = name.trim().lowercase()
        if (normalized.isEmpty()) return null
        return allAreasLocal().firstOrNull { it.areaName.trim().lowercase() == normalized }
    }

    suspend fun createArea(name: String, bazamId: String = kDefaultBazamId): Area {
        findAreaByName(name)?.let { return it }
        val id = UUID.randomUUID().toString()
        val area = Area(
            areaId = id,
            areaName = name.trim(),
            bazamId = bazamId.trim().ifEmpty { kDefaultBazamId },
        )
        areaDao.upsert(area.toEntity(SyncStatus.pendingCreate))
        syncQueueDao.enqueueSync("areas", "create", id, area.toMap())
        return area
    }

    suspend fun upsertAreaLocal(area: Area, syncStatus: String = SyncStatus.synced) {
        areaDao.upsert(area.toEntity(syncStatus))
    }

    suspend fun upsertBazamLocal(bazam: Bazam, syncStatus: String = SyncStatus.synced) {
        bazamDao.upsert(bazam.toEntity(syncStatus))
    }

    private suspend fun allAreasLocal(): List<Area> {
        val rows = areaDao.getAll()
        if (rows.isEmpty()) return kAreas
        return rows
            .filter {
                it.syncStatus != SyncStatus.pendingDelete &&
                    it.syncStatus != SyncStatus.alias
            }
            .map { it.toDomain() }
    }
}
